### PR TITLE
Add global_state_conditions handling

### LINKED ISSUE #62446: [FEATURE REQUEST] Ignore service state actions when in chroot
**Is your feature request related to a problem? Please describe.**
We have a common workflow where we need to manage services, but only when Salt is not running in a chroot. Unless we gate the state blocks with Jinja, it runs and fails since the chroot does not have enough access to perform the action.

**Describe the solution you'd like**
Simple parameter for every state module function which will ignore actions while in chroot, but act normally when outside of it.

**Describe alternatives you've considered**
Wrapper Jinja around _every_ service state block (or macros):

```
{% if grains.get("virtual_subtype") != "chroot" -%}
manage_service:
  service.running:
    - name: service_name

{% endif -%}
```

**Additional context**
Somewhat related to issue #38751, although we **don't** want to take action with this workflow... whereas the solution to that issue would allow some actions to be be taken from within the chroot if allowed.

**Please Note**
If this feature request would be considered a substantial change or addition, this should go through a SEP process here https://github.com/saltstack/salt-enhancement-proposals, instead of a feature request.

### PR DESCRIPTION
**Pull request by nicholasmhughes**
### What does this PR do?
This PR introduces a new minion configuration option which allows for global conditions to be introduced into the state system logic. This should help limit repetitive templating wrappers which are used to enable/disable state blocks in multiple formulas.

### What issues does this PR fix or reference?
Fixes: #62446

### Previous Behavior
If we want to check a standard condition in Grains, we'd have to template around _every_ state block in which we'd want to perform that check:
```
{% if grains.get("virtual_subtype") != "chroot" -%}
manage_service:
  service.running:
    - name: service_name

{% endif -%}
```

### New Behavior
Now we only need to set a minion configuration option on the host:
```
global_state_conditions:
  service: ["not G@virtual_subtype:chroot"]
```
...and then a standard state block such as:
```
manage_service:
  service.running:
    - name: service_name
```
...will not be run on a host which doesn't meet the conditions for the state to be run:
```
----------
          ID: manage_service
    Function: service.running
        Name: service_name
      Result: None
     Comment: Failed to meet global state conditions. State not called.
     Started: 13:30:47.823249
    Duration: 3.012 ms
     Changes:
```

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**

- [x] Docs
- [x] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
Yes

Please review [Salt's Contributing Guide](https://docs.saltproject.io/en/master/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

### CODE REVIEW

**Comment by nicholasmhughes** [comment]
re-run pr-centosstream-9-x86_64-py3-pytest

**Comment by nicholasmhughes** [comment]
re-run pr-freebsd-131-amd64-py3-pytest

**Comment by nicholasmhughes** [comment]
re-run pr-amazon-2-x86_64-py3-pytest

**Comment by nicholasmhughes** [comment]
re-run pr-windows-2019-x64-py3-pytest

**Review by Ch3LL — DISMISSED** [review]

**Comment by Ch3LL** [comment]
Just one question, does this impact performance dramatically when this feature is enabled?

**Comment by nicholasmhughes** [comment]
Once #62295 is merged, this logic will only add a very small amount of processing time to each state block evaluation , but it should be negligible even at scale (+ ~10sec per 100,000 state blocks)... unless someone decides to abuse the heck out of it and have a ridiculous amount of conditions to evaluate.

**Review by whytewolf — COMMENTED** [review]
how about calling config.get module instead. that way the configuration can come from multiple points and not need to be limited to a config option only. I'm also not a fan of config file changes. as to make changes to them a restart of the minion needs to happen. this can be disruptive to users environments.

File: salt/state.py
Comment (nicholasmhughes) [inline salt/state.py:806]:
```suggestion
        if isinstance(self.functions["config.get"]("global_state_conditions"), dict):
```

@whytewolf , like this?

**Comment by nicholasmhughes** [comment]
re-run pr-centosstream-9-x86_64-py3-pytest

File: salt/state.py
Comment (whytewolf) [reply salt/state.py:806]: > 

That is closer. I meant `config.option`, I confuse it with `config.get` a few times, which i know is available within the functions dict.  

also i would suggest dumping the output into it's own variable and passing that variable to the if and for loops. just to save the processing of calling config.option multiple times on the same information. i know the calls are minimal but the ocd in me.

**Comment by nicholasmhughes** [comment]
re-run pr-macosx-catalina-x86_64-py3-pytest

**Comment by nicholasmhughes** [comment]
re-run pr-macosx-catalina-x86_64-py3-pytest

**Review by whytewolf — DISMISSED** [review]

**Review by whytewolf — DISMISSED** [review]

**Comment by nicholasmhughes** [comment]
re-run pr-centos-7-x86_64-py3-tcp-pytest

**Comment by nicholasmhughes** [comment]
re-run pr-alma-8-x86_64-py3-pytest

**Comment by nicholasmhughes** [comment]
re-run pr-macosx-catalina-x86_64-py3-pytest

**Comment by Ch3LL** [comment]
I'm starting to think that mac test failure is related because its failed on every test run on this PR

**Comment by nicholasmhughes** [comment]
well... only test runs since I moved to `config.option`. still not sure what's going on here though... every other platform passes this test in under 30 secs. MacOS already has a 15 sec addition onto that limit in the test code, and it keeps taking 300+ sec in recent runs.

**Comment by nicholasmhughes** [comment]
re-run pr-windows-2019-x64-py3-pytest

File: doc/ref/configuration/minion.rst
Comment (twangboy) [inline doc/ref/configuration/minion.rst:2458]: `and a`

**Review by twangboy — CHANGES_REQUESTED** [review]

**Comment by nicholasmhughes** [comment]
@MKLeb @garethgreenaway , should #62901 have resolved the `pr-macosx-catalina-x86_64-py3-pytest` failure I'm still seeing here?

**Comment by garethgreenaway** [comment]
@nicholasmhughes Doesn't look like the test that failed was the ones addressed in that referenced PR.

**Comment by MKLeb** [comment]
@nicholasmhughes The mac test is not one of the ones fixed by my PR from last week.  I have seen that specific test fail a couple times before, but I can't find an example of it.  However, it does seem to be failing in a rather profound fashion (almost 8 times the expected duration), which has me inclined to agree with @Ch3LL 's comment.  Going to re-run it here and see what happens.

**Comment by nicholasmhughes** [comment]
I just don't get what's so different about MacOS that it's taking that long while every other OS takes under 30 seconds...

File: doc/ref/configuration/minion.rst
Comment (twangboy) [reply doc/ref/configuration/minion.rst]: I took care of it

**Review by twangboy — DISMISSED** [review]

**Comment by MKLeb** [comment]
@nicholasmhughes Are you able to test this on a mac?  I'm wondering if debug statements in your code changes will show this test touching your changes unexpectedly.  If you don't have a mac, I can try to test it out.

**Comment by nicholasmhughes** [comment]
@MKLeb nope. I don't have a Mac.

**Comment by Ch3LL** [comment]
Mac tests should be fixed since this is closed: https://github.com/saltstack/salt/issues/62829  Will update the branch.

**Comment by Ch3LL** [comment]
Can you resolve the merge conflict

**Review by cmcmarrow — DISMISSED** [review]

**Comment by MKLeb** [comment]
@nicholasmhughes Okay, so I've found what is causing the drastic failures.  The test in question is actually touching your changes once (and seemingly _only_ once).  On average, with my testing machine, the method you added is taking ~170ms.  Specifically, the `self.functions["config.option"]` call is taking greater than 95% of that time.  If I take your change out, it returns to passing.  It seems the rest of the state call takes around 20ms (i.e. under the threshold of the test).  Ideally, we'd like to find a way to have your changes coexist in a way which doesn't exceed the remaining 25 seconds of the threshold.

**Comment by nicholasmhughes** [comment]
@MKLeb , we suspected `config.option` as the problem:

> well... only test runs since I moved to `config.option`. still not sure what's going on here though... every other platform passes this test in under 30 secs. MacOS already has a 15 sec addition onto that limit in the test code, and it keeps taking 300+ sec in recent runs.

Not sure what your references to millisecond return times have to do with the "spawning platform" tests that are failing for MacOS, but I still don't understand what _in particular_ is happening inside `config.option` **only** on Macs that is adding that much more time (230-300 sec) that it's failing to return within the 45 second threshold while every other platform returns in under 30 seconds.

Were you able to narrow the "expensive" operations _within_ `config.option` so we might have an idea as to what's happening on Mac?

**Comment by MKLeb** [comment]
Ok... So I don't know _precisely_ what is going on, but I do know one thing for sure - Loading `config.option`, or any `config.<method>`, in the separate process spun up by `call_parallel` takes an _aggressively_ long time.  I pinpointed it right to the line that runs `load_module`.  I wasn't sure how to solve that, so I went with a different approach.  Factoring your call to `_match_global_state_conditions` in `_call_parallel_target` out to the `call` method like how you call the other one allows the test to pass.  Also, I ran your tests with my change and they passed as well.  I can make a suggestion here so it's more clear.

File: salt/state.py
Comment (MKLeb) [inline salt/state.py:2099]:
```suggestion
            ret = instance.states[cdata["full"]](*cdata["args"], **cdata["kwargs"])
```

**Comment by nicholasmhughes** [comment]
Made a change that was close to what you proposed. Just moved the call to `_match_global_state_conditions` outside of the nested `if` since the refactor would've put it in both conditions. Hopefully the tests all run clean now...

**Review by Ch3LL — APPROVED** [review]

**Comment by OrangeDog** [comment]
Any reasons why this was merged with no discussion, when all the comments on the feature request were against it?

**Comment by nicholasmhughes** [comment]
The original implementation outlined in the feature request was not put in place. This PR is a broader feature that was discussed in a couple Open Hours.

**Comment by OrangeDog** [comment]
Ah, ok.

Open Hour discussions should be captured properly. I don't even recall any mention of this in the minutes,

**Pull request merged**
